Implement rmw_test_fixture to start the Zenoh router

cottsay opened the pull request:
It would probably be a good idea to merge the changes to `ZENOH_CONFIG_OVERRIDE` with any existing values, and also to restore the original state when the fixture stops.

Requires:
- ros2/ament_cmake_ros#21
- ros2/ros2#1666
- ros2/ci#810

cottsay committed:
WIP: Implent rmw_test_fixture to start the Zenoh router

Signed-off-by: Scott K Logan <logans@cottsay.net>

cottsay committed:
Proper arbitrary port allocation for the test fixture

It's unfortunate that we need to take a dependency to parse the JSON
string coming out of the session, but this strategy should reliably open
start the router on an arbitrary unused port for the isolated tests to
use.

Yadunund commented on rmw_zenoh_cpp/src/rmw_test_fixture.cpp:
```diff
@@ -0,0 +1,105 @@
+#include <rmw_test_fixture/rmw_test_fixture.h>
+
+#include <cstdlib>
+#include <memory>
```
For std::cerr

```suggestion
#include <iostream>
#include <memory>
```

cottsay replied:
Thanks, done in 8a7be3f

Yadunund commented on rmw_zenoh_cpp/src/rmw_test_fixture.cpp:
```diff
@@ -0,0 +1,105 @@
+std::string
+get_endpoints(zenoh::Session & session)
+{
+  auto zid = session.get_zid();
```
Nit: Similarly avoid auto in other places below.
```suggestion
  const zenoh::Id zid = session.get_zid();
```

cottsay replied:
Thanks, done in 8a7be3f

Yadunund commented on rmw_zenoh_cpp/src/rmw_test_fixture.cpp:
```diff
@@ -0,0 +1,105 @@
+get_endpoints(zenoh::Session & session)
+{
+  auto zid = session.get_zid();
+  auto keyexpr = zenoh::KeyExpr("@/" + zid.to_string() + "/router");
```
```suggestion
  const zenoh::KeyExpr keyexpr = zenoh::KeyExpr("@/" + zid.to_string() + "/router");
```

cottsay replied:
Thanks, done in 8a7be3f

Yadunund commented on rmw_zenoh_cpp/src/rmw_test_fixture.cpp:
```diff
@@ -0,0 +1,105 @@
+static std::unique_ptr<zenoh::Session> g_session = nullptr;
+
+static
+std::string
```
Might be good to return an `std::nullopt if `get` fails
(and include <optional>)
```suggestion
std::optional<std::string>
```

cottsay replied:
Thanks, done in 8a7be3f

Yadunund commented on rmw_zenoh_cpp/src/rmw_test_fixture.cpp:
```diff
@@ -0,0 +1,105 @@
+{
+  auto zid = session.get_zid();
+  auto keyexpr = zenoh::KeyExpr("@/" + zid.to_string() + "/router");
+  auto replies = session.get(keyexpr, "", zenoh::channels::FifoChannel(2));
```
```suggestion
  zenoh::ZResult result
  auto replies = session.get(
    keyexpr,
    "",
    zenoh::channels::FifoChannel(2),
    zenoh::GetOptions()::create_default(),
    &result
  );
  if (result != Z_OK) {
    std::cerr << "Error calling get over keyexpr " << std::string(keyexpr.value().as_string_view()) << std::endl;
    return std::nullopt;
  }
```

cottsay replied:
Thanks, done in 8a7be3f

Yadunund commented on rmw_zenoh_cpp/src/rmw_test_fixture.cpp:
```diff
@@ -0,0 +1,105 @@
+    return RMW_RET_ERROR;
+  }
+
+  std::string endpoints = get_endpoints(*g_session);
+
+  std::string config_override = "connect/endpoints=" + endpoints;
```
```suggestion
  std::optional<std::string> endpoints = get_endpoints(*g_session);
  if (!endpoints.has_value()) {
    // Error already logged.
    return RMW_RET_ERROR;
  }
  const std::string config_override = "connect/endpoints=" + endpoints.value();
```

cottsay replied:
Thanks, done in 8a7be3f

Yadunund commented on rmw_zenoh_cpp/src/rmw_test_fixture.cpp:
```diff
@@ -0,0 +1,105 @@
+extern "C"
+{
+rmw_ret_t
+rmw_test_isolation_start()
```
Could we add some comments to document our approach here? ie Starting a router while having Zenoh assign a port, retrieving the list of locators that the Zenoh router is listening on by querying the Zenoh admin space and finally overriding the `connect` endpoints with the list of locators via the `ZENOH_CONFIG_OVERRIDE` envar?

cottsay replied:
Thanks, done in 8a7be3f

Yadunund commented on rmw_zenoh_cpp/src/rmw_test_fixture.cpp:
```diff
@@ -0,0 +1,105 @@
+
+  auto reply = replies.recv();
+  const auto & sample = std::get<zenoh::Reply>(reply).get_ok();
+  auto parsed = nlohmann::json::parse(sample.get_payload().as_string());
```
Wrap in try-catch block to catch parse errors?

cottsay replied:
Thanks, done in 8a7be3f

Yadunund reviewed (changes requested):
This is great! 

It would be great to also check the result of the `get()` call so i've suggestion some changes around that.

ahcorde commented on rmw_zenoh_cpp/src/rmw_test_fixture.cpp:
```diff
@@ -0,0 +1,105 @@
+static std::unique_ptr<zenoh::Session> g_session = nullptr;
+
+static
+std::string
```
include `<string>`

cottsay replied:
Thanks, done in 8a7be3f

ahcorde commented on rmw_zenoh_cpp/src/rmw_test_fixture.cpp:
```diff
@@ -0,0 +1,105 @@
+  }
+
+  g_session = std::make_unique<zenoh::Session>(
+    std::move(config.value()),
```
include `<utility>`

cottsay replied:
Thanks, done in 8a7be3f

cottsay committed:
Merge branch 'rolling' into cottsay/rmw_test_fixture

cottsay committed:
Stop connecting to an upstream router

If the Zenoh config specifies a connect/endpoints value, ignore it
during isolated testing. The user can disable isolation if they actually
want the tests to connect to a specific endpoint.

cottsay committed:
PR feedback: error handling, includes, etc

Signed-off-by: Scott K Logan <logans@cottsay.net>

Yadunund reviewed (approved):
LGTM. Thanks for iterating on the PR so quickly!

cottsay commented:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22973)](http://ci.ros2.org/job/ci_linux/22973/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=17194)](http://ci.ros2.org/job/ci_linux-aarch64/17194/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=2460)](http://ci.ros2.org/job/ci_linux-rhel/2460/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23849)](http://ci.ros2.org/job/ci_windows/23849/)

ahcorde commented:
Pulls: ros2/rmw_zenoh#583
Gist: https://gist.githubusercontent.com/ahcorde/e9bfa8ceb23242512089e1f8dba249d6/raw/2bf4195990589fbfc38e384a0f2b23267b522362/ros2.repos
BUILD args:  --packages-up-to rmw_zenoh_cpp  --packages-above-and-dependencies rmw_zenoh_cpp
TEST args:  --packages-select rmw_zenoh_cpp  --packages-above rmw_zenoh_cpp
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/15548
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22974)](http://ci.ros2.org/job/ci_linux/22974/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=17195)](http://ci.ros2.org/job/ci_linux-aarch64/17195/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=2461)](http://ci.ros2.org/job/ci_linux-rhel/2461/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23850)](http://ci.ros2.org/job/ci_windows/23850/)

cottsay committed:
Explicitly disable multicast discovery during testing

cottsay commented:
CI was failing due to a branch name collision in rcutils. I cleaned that up and we should be in better shape now:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=22982)](http://ci.ros2.org/job/ci_linux/22982/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=17205)](http://ci.ros2.org/job/ci_linux-aarch64/17205/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=2471)](http://ci.ros2.org/job/ci_linux-rhel/2471/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23858)](http://ci.ros2.org/job/ci_windows/23858/)

Still expecting the binary PR build to fail because debs for `rcutils` and `rmw_test_fixture` haven't been synced yet.

cottsay commented:
Ah, rats, I need to get the new dependency into the CI builds before this can be merged.

cottsay committed:
Set missing DLL export flag for Windows

cottsay commented:
Fresh builds :crossed_fingers::
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=23006)](http://ci.ros2.org/job/ci_linux/23006/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=17232)](http://ci.ros2.org/job/ci_linux-aarch64/17232/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=2493)](http://ci.ros2.org/job/ci_linux-rhel/2493/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=23881)](http://ci.ros2.org/job/ci_windows/23881/)
